3단계 - 자동차 경주

traumereiii opened the pull request:
안녕하세요 리뷰어님
이번에도 리뷰 잘 부탁드립니다 :)

DongSeong-Seo committed:
chore: ktlint 수정

DongSeong-Seo committed:
feat: 자동차 게임 구현

ohgillwhan commented on src/main/kotlin/study/step3/domain/Car.kt:
```diff
@@ -0,0 +1,10 @@
+package study.step3.domain
+
+class Car(private val engine: CarEngine) {
+    var position: Int = 1
+        private set
```
외부에서 `getter` 호출이 가능해 보이네요! 함수의 리턴을 잘 활용하면 `getter`도 없앨 수 있을거에요.

ohgillwhan commented on src/main/kotlin/study/step3/domain/Car.kt:
```diff
@@ -0,0 +1,10 @@
+package study.step3.domain
+
+class Car(private val engine: CarEngine) {
```
전략패턴 💯

ohgillwhan commented on src/main/kotlin/study/step3/ui/RacingGamePrinter.kt:
```diff
@@ -0,0 +1,16 @@
+class CommandLineRacingGamePrinter : RacingGamePrinter {
+    override fun print(cars: List<Car>) {
+        cars.forEach {
+            (0 until it.position).forEach { position -> print('-') }
+            println()
```
`println("-".repeat(it.position))`도 가능할거 같아요

traumereiii replied:
Car의 position을 알기 위해선 getter 혹은 이와 비슷한 함수가 필요할 것으로 생각되는데 
getter 말고 함수를 정의하는게 좋을까요? 
단순히 position을 반환하는 함수라면 getter와 다를게 없다 생각해서요!

lullaaaby13 replied:
생각나는 다른 방법으로는 CarPrinter라는 인터페이스를 Car의 메서드로 전달하여 Car가 직접 출력하도록 하는 방법도 가능할 것 같은데
이런 경우에 대해선 어떻게 생각하시는지 궁금합니다!

ohgillwhan commented on src/main/kotlin/study/step3/domain/RacingGame.kt:
```diff
@@ -0,0 +1,23 @@
+
+    companion object {
+        fun of(racingGameConfiguration: RacingGameConfiguration): RacingGame {
+            val cars = (0 until racingGameConfiguration.numberOfCars).map { Car(RandomSeedEngine()) }
```
해당 코드에서 차를 만드는게 아니라, `RacingGameConfiguration`에서 차를 만들어주는건 어떨까요?

ohgillwhan replied:
더 나아가서 `RacingGameConfiguration`에서 `RacingGame` 자체를 만들어줘도 괜찮아 보이네요.
그러면 `RacingGame`은 `RacingGameConfiguration`에 의존하지 않고, `List<Car>`, `Int`만 알고있으면 될거에요.

ohgillwhan commented on src/main/kotlin/study/step3/Main.kt:
```diff
@@ -0,0 +1,16 @@
+    val racingGameConfigurator = CommandLineRacingGameConfigurator()
+    val racingGameConfiguration = racingGameConfigurator.run()
+
+    val racingGame = RacingGame.of(racingGameConfiguration)
```
`RacingGameConfiguration `에서 `RacingGame`을 만들게 하는건 어떨까요?

ohgillwhan commented on src/main/kotlin/study/step3/domain/RacingGameConfiguration.kt:
```diff
@@ -0,0 +1,19 @@
+package study.step3.domain
+
+class RacingGameConfiguration private constructor(
```
`companion object`를 포함한 모든 함수는 테스트코드가 필요합니다!

ohgillwhan commented on src/main/kotlin/study/step3/domain/RacingGame.kt:
```diff
@@ -0,0 +1,23 @@
+
+import study.step3.ui.RacingGamePrinter
+
+class RacingGame private constructor (
```
`companion object`를 포함한 모든 함수는 테스트코드가 필요합니다!

traumereiii replied:
이번테 테스트를 작성하지 않은 이유는 어떻게 테슨트해야할지 감이 안와서인데요..ㅠ
RacingGame의 start라는 메서드를 실행했을 때
무엇을 검증하는 게 좋을까요?

ohgillwhan commented on src/main/kotlin/study/step3/domain/CarEngine.kt:
```diff
@@ -0,0 +1,25 @@
+    }
+}
+
+class BrokenEngine : CarEngine {
+    override fun run(): Int {
+        return 0
+    }
+}
+
+class NormalEngine : CarEngine {
+    override fun run(): Int {
+        return 1
+    }
+}
```
테스트코드를 위해서 만든것 같네요! 테스트코드 쪽으로 이동해볼까요?

ohgillwhan reviewed (changes requested):
코드 잘 봤습니다! 리뷰 몇가지 남겼으니, 확인 부탁드려요 💯

DongSeong-Seo committed:
feat: #1085 PR 리뷰 반영

- Car.position 외부에서 참조하지 못하도록 변경
- Car 테스트를 위한 CarEngine 구현체를 테스트쪽으로 이동
- Car의 position을 출력하기 위한 인터페이스 추가
- RacingGame 생성을 RacingGameConfiguration에게 위임
- RacingGameConfiguration 테스트 추가

ohgillwhan commented on src/main/kotlin/study/step3/Main.kt:
```diff
@@ -0,0 +1,15 @@
+
+    println("실행 결과")
+
+    racingGame.start(CommandLineRacingGamePrinter)
```
도메인이 `ui`를 알지 않아도 될거 같아요.

ohgillwhan commented on src/main/kotlin/study/step3/domain/Car.kt:
```diff
@@ -0,0 +1,16 @@
+        return position
+    }
+
+    fun printPosition(printer: CarPositionPrinter) {
+        printer.print(position)
+    }
```
도메인은 입/출력을 알면 안될것 같아요.

ohgillwhan commented on src/main/kotlin/study/step3/domain/RacingGame.kt:
```diff
@@ -0,0 +1,16 @@
+    val numberOfTries: Int,
+) {
+
+    fun start(racingGamePrinter: RacingGamePrinter) {
```
이 부분에서도 출력을 제거해볼까요?

ohgillwhan replied:
리턴을 잘 활용하면 충분히 될거 같네요!

ohgillwhan reviewed (approved):
코드 💯 
도메인이 `ui`를 알지 못하도록 다음 `step`에서 진행해볼까요?
